applications/nrf5340_audio/src/bluetooth/bt_management/bt_mgmt_scan.c

```diff
@@ -0,0 +1,287 @@
+			 CONFIG_BLE_ACL_SLAVE_LATENCY, CONFIG_BLE_ACL_SUP_TIMEOUT)
+
+static bt_mgmt_conn_set_cb conn_set;
+static bt_mgmt_conn_disconnected_cb conn_disconnected;
+
+static int scan_interval_ms;
+static int scan_window_ms;
+
+static uint8_t num_con;
+static uint8_t bonded_num;
+
+static void bond_check(const struct bt_bond_info *info, void *user_data)
+{
+	char addr_buf[BT_ADDR_LE_STR_LEN];
+
+	bt_addr_le_to_str(&info->addr, addr_buf, BT_ADDR_LE_STR_LEN);
+
+	LOG_DBG("Stored bonding found: %s", addr_buf);
+	bonded_num++;
+}
+
+static void bond_connect(const struct bt_bond_info *info, void *user_data)
+{
+	int ret;
+	const bt_addr_le_t *adv_addr = user_data;
+	struct bt_conn *conn;
+
+	if (!bt_addr_le_cmp(&info->addr, adv_addr)) {
+		LOG_DBG("Found bonded device");
+
+		ret = bt_le_scan_stop();
+		if (ret) {
+			LOG_WRN("Stop scan failed: %d", ret);
+		}
+
+		ret = bt_conn_le_create(adv_addr, BT_CONN_LE_CREATE_CONN, CONNECTION_PARAMETERS,
+					&conn);
+		if (ret) {
+			LOG_WRN("Create ACL connection failed: %d", ret);
+			bt_mgmt_scan_start();
+		}
+	}
+}
+
+static void connected_cb(struct bt_conn *conn, uint8_t err)
+{
+	int ret;
+	char addr[BT_ADDR_LE_STR_LEN];
+	uint16_t conn_handle;
+	enum ble_hci_vs_tx_power conn_tx_pwr;
+
+	(void)bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	if (err) {
+		LOG_ERR("ACL connection to %s failed, error %d", addr, err);
+
+		bt_conn_unref(conn);
+		bt_mgmt_scan_start();
+
+		return;
+	}
+
+	num_con++;
+
+	/* ACL connection established */
+	LOG_INF("Connected: %s", addr);
+
+	ret = bt_hci_get_conn_handle(conn, &conn_handle);
+	if (ret) {
+		LOG_ERR("Unable to get conn handle");
+	} else {
+#if (CONFIG_NRF_21540_ACTIVE)
+		conn_tx_pwr = CONFIG_NRF_21540_MAIN_DBM;
+#else
+		conn_tx_pwr = CONFIG_BLE_CONN_TX_POWER_DBM;
+#endif /* (CONFIG_NRF_21540_ACTIVE) */
+		ret = ble_hci_vsc_conn_tx_pwr_set(conn_handle, conn_tx_pwr);
+		if (ret) {
+			LOG_ERR("Failed to set TX power for conn");
+		} else {
+			LOG_DBG("TX power set to %d dBm for connection %p", conn_tx_pwr,
+				(void *)conn);
+		}
+	}
+
+	ret = bt_conn_set_security(conn, BT_SECURITY_L2);
+	if (ret) {
+		LOG_ERR("Failed to set security to L2: %d", ret);
+	}
+}
+
+static void disconnected_cb(struct bt_conn *conn, uint8_t reason)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	(void)bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	LOG_INF("Disconnected: %s (reason 0x%02x)", addr, reason);
+
+	bt_conn_unref(conn);
+
+	conn_disconnected(conn);
+
+	num_con--;
+
+	bt_mgmt_scan_start();
+}
+
+static void security_changed_cb(struct bt_conn *conn, bt_security_t level, enum bt_security_err err)
+{
+	int ret;
+
+	if (err) {
+		LOG_ERR("Security failed: level %d err %d", level, err);
+		ret = bt_conn_disconnect(conn, err);
+		if (ret) {
+			LOG_ERR("Failed to disconnect %d", ret);
+		}
+	} else {
+		LOG_DBG("Security changed: level %d", level);
+		conn_set(conn);
```

Comment:
This is where you could check num_conn against CONFIG_BT_MAX_CONN and start_scan again if not equal

Reply:
I have placed this functionality in connected_cb now.
Maybe it is better to have it in security_changed_cb() to avoid stressing the controller?

Reply:
Yeah, connected_cb happens before this, and there are still quite a few packets that needs to go over the air, so I think it's better to wait as long as possible

Reply:
Hmm never mind my comment, conn_set(conn) is already in security_changed_cb()